File: runtime/go/frame.go
// Package ridlruntime implements the transport-neutral RIDL frame envelope.
//
// It is a standard-library-only port of ridl/framing.py. HTTP does not use this
// envelope; WebSocket carries one encoded frame per text message and TCP uses a
// four-byte big-endian length prefix followed by the same UTF-8 JSON bytes.
package ridlruntime

import (
	"bytes"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"sort"
	"strconv"
	"strings"
	"sync/atomic"
	"unicode/utf8"
)

const (
	FrameVersion      uint8 = 1
	MaxFrameBytes           = 8 * 1024 * 1024
	LengthPrefixBytes       = 4
)

type FrameKind string

const (
	Call   FrameKind = "call"
	Data   FrameKind = "data"
	End    FrameKind = "end"
	Error  FrameKind = "error"
	Cancel FrameKind = "cancel"
)

var allowedFields = map[string]struct{}{
	"v": {}, "id": {}, "t": {}, "key": {}, "method": {}, "path": {},
	"query": {}, "body": {}, "code": {}, "message": {}, "meta": {},
}

type QueryPair [2]string

// Frame distinguishes an absent body (HasBody=false) from a JSON null body
// (HasBody=true and Body == "null"). Body is raw JSON so object member order is
// preserved across decode/encode; Encode compacts it and rejects invalid JSON.
type Frame struct {
	Version uint8
	ID      string
	Kind    FrameKind
	Key     string
	Method  string
	Path    string
	Query   []QueryPair
	Body    json.RawMessage
	HasBody bool
	Code    string
	Message *string
	Meta    map[string]string
}

func CallFrame(id, key, method, path string, query []QueryPair, body json.RawMessage) Frame {
	return Frame{
		Version: FrameVersion,
		ID:      id,
		Kind:    Call,
		Key:     key,
		Method:  method,
		Path:    path,
		Query:   append([]QueryPair(nil), query...),
		Body:    cloneBytes(body),
		HasBody: body != nil,
		Meta:    map[string]string{},
	}
}

func DataFrame(id string, body json.RawMessage) Frame {
	return Frame{Version: FrameVersion, ID: id, Kind: Data, Body: cloneBytes(body), HasBody: true, Meta: map[string]string{}}
}

func EndFrame(id string) Frame {
	return Frame{Version: FrameVersion, ID: id, Kind: End, Meta: map[string]string{}}
}

func CancelFrame(id string) Frame {
	return Frame{Version: FrameVersion, ID: id, Kind: Cancel, Meta: map[string]string{}}
}

func ErrorFrame(id, code string, message *string, body json.RawMessage) Frame {
	return Frame{Version: FrameVersion, ID: id, Kind: Error, Code: code, Message: message, Body: cloneBytes(body), HasBody: body != nil, Meta: map[string]string{}}
}

func (f Frame) WithMeta(name, value string) Frame {
	copied := make(map[string]string, len(f.Meta)+1)
	for key, item := range f.Meta {
		copied[key] = item
	}
	copied[name] = value
	f.Meta = copied
	return f
}

func (f Frame) Validate() error {
	if f.Version != FrameVersion {
		return fmt.Errorf("unsupported frame version %d", f.Version)
	}
	if !utf8.ValidString(f.ID) || f.ID == "" || utf8.RuneCountInString(f.ID) > 128 {
		return errors.New("id must be 1..128 characters")
	}
	switch f.Kind {
	case Call, Data, End, Error, Cancel:
	default:
		return fmt.Errorf("unknown frame type %q", f.Kind)
	}
	if f.Kind == Call {
		if f.Key == "" {
			return errors.New("a call frame needs an operation key")
		}
		if f.Method == "" {
			return errors.New("a call frame needs a method")
		}
		if !strings.HasPrefix(f.Path, "/") {
			return errors.New("a call frame needs a path starting with /")
		}
	} else if f.Key != "" || f.Method != "" || f.Path != "" || len(f.Query) != 0 {
		return fmt.Errorf("a %s frame carries no addressing fields", f.Kind)
	}
	if f.Kind == Data && !f.HasBody {
		return errors.New("a data frame needs a body")
	}
	if f.Kind == Error {
		if f.Code == "" {
			return errors.New("an error frame needs a code")
		}
	} else if f.Code != "" || f.Message != nil {
		return fmt.Errorf("a %s frame carries no code or message", f.Kind)
	}
	if f.HasBody {
		if len(f.Body) == 0 || !json.Valid(f.Body) {
			return errors.New("body must contain one valid JSON value")
		}
	} else if len(f.Body) != 0 {
		return errors.New("body bytes are present while HasBody is false")
	}
	for index, pair := range f.Query {
		if !utf8.ValidString(pair[0]) || !utf8.ValidString(pair[1]) {
			return fmt.Errorf("query[%d] must contain UTF-8 strings", index)
		}
	}
	for key, value := range f.Meta {
		if !utf8.ValidString(key) || !utf8.ValidString(value) {
			return fmt.Errorf("meta.%s must be a UTF-8 string", key)
		}
	}
	return nil
}

func (f Frame) Encode() ([]byte, error) {
	if err := f.Validate(); err != nil {
		return nil, err
	}
	var out bytes.Buffer
	out.Grow(256)
	out.WriteString(`{"v":`)
	out.WriteString(strconv.Itoa(int(f.Version)))
	out.WriteString(`,"id":`)
	if err := writeJSONString(&out, f.ID); err != nil {
		return nil, err
	}
	out.WriteString(`,"t":`)
	if err := writeJSONString(&out, string(f.Kind)); err != nil {
		return nil, err
	}
	if f.Kind == Call {
		out.WriteString(`,"key":`)
		if err := writeJSONString(&out, f.Key); err != nil {
			return nil, err
		}
		out.WriteString(`,"method":`)
		if err := writeJSONString(&out, f.Method); err != nil {
			return nil, err
		}
		out.WriteString(`,"path":`)
		if err := writeJSONString(&out, f.Path); err != nil {
			return nil, err
		}
		if len(f.Query) > 0 {
			out.WriteString(`,"query":[`)
			for index, pair := range f.Query {
				if index > 0 {
					out.WriteByte(',')
				}
				out.WriteByte('[')
				if err := writeJSONString(&out, pair[0]); err != nil {
					return nil, err
				}
				out.WriteByte(',')
				if err := writeJSONString(&out, pair[1]); err != nil {
					return nil, err
				}
				out.WriteByte(']')
			}
			out.WriteByte(']')
		}
	}
	if f.HasBody {
		out.WriteString(`,"body":`)
		if err := compactJSON(&out, f.Body); err != nil {
			return nil, err
		}
	}
	if f.Kind == Error {
		out.WriteString(`,"code":`)
		if err := writeJSONString(&out, f.Code); err != nil {
			return nil, err
		}
		if f.Message != nil {
			out.WriteString(`,"message":`)
			if err := writeJSONString(&out, *f.Message); err != nil {
				return nil, err
			}
		}
	}
	if len(f.Meta) > 0 {
		keys := make([]string, 0, len(f.Meta))
		for key := range f.Meta {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		out.WriteString(`,"meta":{`)
		for index, key := range keys {
			if index > 0 {
				out.WriteByte(',')
			}
			if err := writeJSONString(&out, key); err != nil {
				return nil, err
			}
			out.WriteByte(':')
			if err := writeJSONString(&out, f.Meta[key]); err != nil {
				return nil, err
			}
		}
		out.WriteByte('}')
	}
	out.WriteByte('}')
	if out.Len() > MaxFrameBytes {
		return nil, fmt.Errorf("frame is %d bytes, over the %d limit", out.Len(), MaxFrameBytes)
	}
	return out.Bytes(), nil
}

func (f Frame) EncodeTCP() ([]byte, error) {
	payload, err := f.Encode()
	if err != nil {
		return nil, err
	}
	out := make([]byte, LengthPrefixBytes+len(payload))
	binary.BigEndian.PutUint32(out[:LengthPrefixBytes], uint32(len(payload)))
	copy(out[LengthPrefixBytes:], payload)
	return out, nil
}

func Decode(payload []byte) (Frame, error) {
	if len(payload) > MaxFrameBytes {
		return Frame{}, fmt.Errorf("frame is %d bytes, over the %d limit", len(payload), MaxFrameBytes)
	}
	if !utf8.Valid(payload) {
		return Frame{}, errors.New("frame is not UTF-8")
	}
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(payload, &raw); err != nil {
		return Frame{}, fmt.Errorf("frame is not JSON: %w", err)
	}
	if raw == nil {
		return Frame{}, errors.New("a frame must be a JSON object")
	}
	unknown := make([]string, 0)
	for key := range raw {
		if _, ok := allowedFields[key]; !ok {
			unknown = append(unknown, key)
		}
	}
	if len(unknown) > 0 {
		sort.Strings(unknown)
		return Frame{}, fmt.Errorf("unknown frame member(s): %s", strings.Join(unknown, ", "))
	}

	var frame Frame
	frame.Meta = map[string]string{}
	if err := required(raw, "v", &frame.Version); err != nil {
		return Frame{}, err
	}
	if err := required(raw, "id", &frame.ID); err != nil {
		return Frame{}, err
	}
	var kind string
	if err := required(raw, "t", &kind); err != nil {
		return Frame{}, err
	}
	frame.Kind = FrameKind(kind)
	if err := optional(raw, "key", &frame.Key); err != nil {
		return Frame{}, err
	}
	if err := optional(raw, "method", &frame.Method); err != nil {
		return Frame{}, err
	}
	if err := optional(raw, "path", &frame.Path); err != nil {
		return Frame{}, err
	}
	if value, ok := raw["query"]; ok {
		var pairs [][]string
		if err := json.Unmarshal(value, &pairs); err != nil {
			return Frame{}, errors.New("query must be an array of [name, value] pairs")
		}
		for _, pair := range pairs {
			if len(pair) != 2 {
				return Frame{}, errors.New("each query entry must be a [name, value] pair of strings")
			}
			frame.Query = append(frame.Query, QueryPair{pair[0], pair[1]})
		}
	}
	if value, ok := raw["body"]; ok {
		frame.HasBody = true
		frame.Body = cloneBytes(value)
	}
	if err := optional(raw, "code", &frame.Code); err != nil {
		return Frame{}, err
	}
	if value, ok := raw["message"]; ok {
		var message string
		if err := json.Unmarshal(value, &message); err != nil {
			return Frame{}, errors.New("message must be a string")
		}
		frame.Message = &message
	}
	if value, ok := raw["meta"]; ok {
		if err := json.Unmarshal(value, &frame.Meta); err != nil {
			return Frame{}, errors.New("meta must be an object of strings")
		}
		if frame.Meta == nil {
			return Frame{}, errors.New("meta must be an object of strings")
		}
	}
	if err := frame.Validate(); err != nil {
		return Frame{}, err
	}
	return frame, nil
}

func DecodeStream(buffer []byte) ([]Frame, []byte, error) {
	frames := make([]Frame, 0)
	offset := 0
	for len(buffer)-offset >= LengthPrefixBytes {
		length := int(binary.BigEndian.Uint32(buffer[offset : offset+LengthPrefixBytes]))
		if length > MaxFrameBytes {
			return nil, nil, fmt.Errorf("declared frame length %d is over the %d limit", length, MaxFrameBytes)
		}
		start := offset + LengthPrefixBytes
		if len(buffer)-start < length {
			break
		}
		frame, err := Decode(buffer[start : start+length])
		if err != nil {
			return nil, nil, err
		}
		frames = append(frames, frame)
		offset = start + length
	}
	return frames, buffer[offset:], nil
}

type Correlator struct {
	prefix string
	next   atomic.Uint64
}

func NewCorrelator(prefix string) *Correlator { return &Correlator{prefix: prefix} }

func (c *Correlator) Take() string {
	value := c.next.Add(1)
	return c.prefix + strconv.FormatUint(value, 10)
}

func writeJSONString(out *bytes.Buffer, value string) error {
	if !utf8.ValidString(value) {
		return errors.New("string is not UTF-8")
	}
	var encoded bytes.Buffer
	encoder := json.NewEncoder(&encoded)
	encoder.SetEscapeHTML(false)
	if err := encoder.Encode(value); err != nil {
		return err
	}
	bytesValue := encoded.Bytes()
	out.Write(bytesValue[:len(bytesValue)-1])
	return nil
}

func compactJSON(out *bytes.Buffer, value json.RawMessage) error {
	if !json.Valid(value) {
		return errors.New("body must contain one valid JSON value")
	}
	var compacted bytes.Buffer
	if err := json.Compact(&compacted, value); err != nil {
		return err
	}
	out.Write(compacted.Bytes())
	return nil
}

func required(raw map[string]json.RawMessage, name string, target any) error {
	value, ok := raw[name]
	if !ok {
		return fmt.Errorf("missing frame member %s", name)
	}
	if err := json.Unmarshal(value, target); err != nil {
		return fmt.Errorf("%s has the wrong type", name)
	}
	return nil
}

func optional(raw map[string]json.RawMessage, name string, target any) error {
	value, ok := raw[name]
	if !ok {
		return nil
	}
	if err := json.Unmarshal(value, target); err != nil {
		return fmt.Errorf("%s has the wrong type", name)
	}
	return nil
}

func cloneBytes(value []byte) []byte {
	if value == nil {
		return nil
	}
	return append([]byte(nil), value...)
}

File: runtime/dart/lib/src/frame.dart
/// Transport-neutral RIDL frame codec for Flutter, browsers, and Dart servers.
///
/// HTTP does not use this envelope. WebSocket carries one encoded frame per
/// text message; TCP carries a four-byte big-endian length followed by the same
/// canonical UTF-8 JSON bytes.
import 'dart:convert';
import 'dart:typed_data';

const int frameVersion = 1;
const int maxFrameBytes = 8 * 1024 * 1024;
const int lengthPrefixBytes = 4;

const Set<String> _allowedFields = {
  'v',
  'id',
  't',
  'key',
  'method',
  'path',
  'query',
  'body',
  'code',
  'message',
  'meta',
};

enum FrameKind { call, data, end, error, cancel }

extension FrameKindWire on FrameKind {
  String get wireName => name;
}

FrameKind _parseFrameKind(String value) => switch (value) {
      'call' => FrameKind.call,
      'data' => FrameKind.data,
      'end' => FrameKind.end,
      'error' => FrameKind.error,
      'cancel' => FrameKind.cancel,
      _ => throw FrameException('unknown frame type $value'),
    };

final class FrameException implements Exception {
  const FrameException(this.message);

  final String message;

  @override
  String toString() => 'FrameException: $message';
}

final class QueryPair {
  const QueryPair(this.name, this.value);

  final String name;
  final String value;
}

/// Wrapper used where JSON `null` must remain distinct from an absent body.
final class PresentBody {
  const PresentBody(this.value);

  final Object? value;
}

final class Frame {
  Frame._({
    required this.version,
    required this.id,
    required this.kind,
    this.key,
    this.method,
    this.path,
    List<QueryPair> query = const [],
    this.body,
    required this.hasBody,
    this.code,
    this.message,
    Map<String, String> meta = const {},
  })  : query = List.unmodifiable(query),
        meta = Map.unmodifiable(meta);

  factory Frame.call({
    required String id,
    required String key,
    required String method,
    required String path,
    List<QueryPair> query = const [],
    PresentBody? body,
    Map<String, String> meta = const {},
  }) =>
      Frame._(
        version: frameVersion,
        id: id,
        kind: FrameKind.call,
        key: key,
        method: method,
        path: path,
        query: query,
        body: body?.value,
        hasBody: body != null,
        meta: meta,
      );

  factory Frame.data({required String id, required Object? body}) => Frame._(
        version: frameVersion,
        id: id,
        kind: FrameKind.data,
        body: body,
        hasBody: true,
      );

  factory Frame.end({required String id}) => Frame._(
        version: frameVersion,
        id: id,
        kind: FrameKind.end,
        hasBody: false,
      );

  factory Frame.cancel({required String id}) => Frame._(
        version: frameVersion,
        id: id,
        kind: FrameKind.cancel,
        hasBody: false,
      );

  factory Frame.error({
    required String id,
    required String code,
    String? message,
    PresentBody? body,
    Map<String, String> meta = const {},
  }) =>
      Frame._(
        version: frameVersion,
        id: id,
        kind: FrameKind.error,
        body: body?.value,
        hasBody: body != null,
        code: code,
        message: message,
        meta: meta,
      );

  final int version;
  final String id;
  final FrameKind kind;
  final String? key;
  final String? method;
  final String? path;
  final List<QueryPair> query;
  final Object? body;
  final bool hasBody;
  final String? code;
  final String? message;
  final Map<String, String> meta;

  Frame withMeta(String name, String value) {
    final next = Map<String, String>.of(meta)..[name] = value;
    return Frame._(
      version: version,
      id: id,
      kind: kind,
      key: key,
      method: method,
      path: path,
      query: query,
      body: body,
      hasBody: hasBody,
      code: code,
      message: message,
      meta: next,
    );
  }

  void validate() {
    if (version != frameVersion) {
      throw FrameException('unsupported frame version $version');
    }
    if (id.isEmpty || id.runes.length > 128 || !_isUnicodeScalarString(id)) {
      throw const FrameException('id must be 1..128 Unicode scalar values');
    }
    if (kind == FrameKind.call) {
      if (key == null || key!.isEmpty || !_isUnicodeScalarString(key!)) {
        throw const FrameException('a call frame needs an operation key');
      }
      if (method == null || method!.isEmpty || !_isUnicodeScalarString(method!)) {
        throw const FrameException('a call frame needs a method');
      }
      if (path == null || !path!.startsWith('/') || !_isUnicodeScalarString(path!)) {
        throw const FrameException('a call frame needs a path starting with /');
      }
    } else if (key != null || method != null || path != null || query.isNotEmpty) {
      throw FrameException('a ${kind.wireName} frame carries no addressing fields');
    }
    for (var index = 0; index < query.length; index += 1) {
      final pair = query[index];
      if (!_isUnicodeScalarString(pair.name) || !_isUnicodeScalarString(pair.value)) {
        throw FrameException('query[$index] must contain Unicode scalar strings');
      }
    }
    if (kind == FrameKind.data && !hasBody) {
      throw const FrameException('a data frame needs a body');
    }
    if (kind == FrameKind.error) {
      if (code == null || code!.isEmpty || !_isUnicodeScalarString(code!)) {
        throw const FrameException('an error frame needs a code');
      }
      if (message != null && !_isUnicodeScalarString(message!)) {
        throw const FrameException('an error message must contain Unicode scalars');
      }
    } else if (code != null || message != null) {
      throw FrameException('a ${kind.wireName} frame carries no code or message');
    }
    if (hasBody && !_isJsonValue(body)) {
      throw const FrameException('body must contain one JSON value');
    }
    for (final entry in meta.entries) {
      if (!_isUnicodeScalarString(entry.key) || !_isUnicodeScalarString(entry.value)) {
        throw FrameException('meta.${entry.key} must contain Unicode scalar strings');
      }
    }
  }

  Map<String, Object?> toCanonicalObject() {
    validate();
    final raw = <String, Object?>{
      'v': version,
      'id': id,
      't': kind.wireName,
    };
    if (kind == FrameKind.call) {
      raw['key'] = key;
      raw['method'] = method;
      raw['path'] = path;
      if (query.isNotEmpty) {
        raw['query'] = query.map((pair) => [pair.name, pair.value]).toList(growable: false);
      }
    }
    if (hasBody) raw['body'] = body;
    if (kind == FrameKind.error) {
      raw['code'] = code;
      if (message != null) raw['message'] = message;
    }
    if (meta.isNotEmpty) {
      final keys = meta.keys.toList()..sort();
      raw['meta'] = <String, String>{for (final key in keys) key: meta[key]!};
    }
    return raw;
  }
}

Uint8List encodeFrame(Frame frame) {
  late final String text;
  try {
    text = jsonEncode(frame.toCanonicalObject());
  } on Object catch (error) {
    if (error is FrameException) rethrow;
    throw FrameException('frame is not JSON-encodable: $error');
  }
  final bytes = Uint8List.fromList(utf8.encode(text));
  if (bytes.length > maxFrameBytes) {
    throw FrameException(
      'frame is ${bytes.length} bytes, over the $maxFrameBytes limit',
    );
  }
  return bytes;
}

Uint8List encodeTcp(Frame frame) {
  final payload = encodeFrame(frame);
  final output = Uint8List(lengthPrefixBytes + payload.length);
  ByteData.sublistView(output).setUint32(0, payload.length, Endian.big);
  output.setRange(lengthPrefixBytes, output.length, payload);
  return output;
}

Frame decodeFrame(Object payload) {
  final String text;
  if (payload is String) {
    final size = utf8.encode(payload).length;
    if (size > maxFrameBytes) {
      throw FrameException('frame is $size bytes, over the $maxFrameBytes limit');
    }
    text = payload;
  } else if (payload is List<int>) {
    if (payload.length > maxFrameBytes) {
      throw FrameException(
        'frame is ${payload.length} bytes, over the $maxFrameBytes limit',
      );
    }
    if (payload.any((value) => value < 0 || value > 255)) {
      throw const FrameException('frame byte values must be between 0 and 255');
    }
    try {
      text = utf8.decode(payload, allowMalformed: false);
    } on FormatException catch (error) {
      throw FrameException('frame is not UTF-8: $error');
    }
  } else {
    throw const FrameException('frame input must be a String or byte list');
  }

  final Object? decoded;
  try {
    decoded = jsonDecode(text);
  } on FormatException catch (error) {
    throw FrameException('frame is not JSON: $error');
  }
  if (decoded is! Map) {
    throw const FrameException('a frame must be a JSON object');
  }
  final raw = <String, Object?>{};
  for (final entry in decoded.entries) {
    if (entry.key is! String) {
      throw const FrameException('frame member names must be strings');
    }
    raw[entry.key as String] = entry.value;
  }
  final unknown = raw.keys.where((key) => !_allowedFields.contains(key)).toList()..sort();
  if (unknown.isNotEmpty) {
    throw FrameException('unknown frame member(s): ${unknown.join(', ')}');
  }

  final version = _requiredInt(raw, 'v');
  final id = _requiredString(raw, 'id');
  final kind = _parseFrameKind(_requiredString(raw, 't'));
  final key = _optionalString(raw, 'key');
  final method = _optionalString(raw, 'method');
  final path = _optionalString(raw, 'path');
  final query = <QueryPair>[];
  if (raw.containsKey('query')) {
    final value = raw['query'];
    if (value is! List) {
      throw const FrameException('query must be an array of [name, value] pairs');
    }
    for (final pair in value) {
      if (pair is! List || pair.length != 2 || pair[0] is! String || pair[1] is! String) {
        throw const FrameException(
          'each query entry must be a [name, value] pair of strings',
        );
      }
      query.add(QueryPair(pair[0] as String, pair[1] as String));
    }
  }

  final meta = <String, String>{};
  if (raw.containsKey('meta')) {
    final value = raw['meta'];
    if (value is! Map) {
      throw const FrameException('meta must be an object of strings');
    }
    for (final entry in value.entries) {
      if (entry.key is! String || entry.value is! String) {
        throw const FrameException('meta must be an object of strings');
      }
      meta[entry.key as String] = entry.value as String;
    }
  }

  final frame = Frame._(
    version: version,
    id: id,
    kind: kind,
    key: key,
    method: method,
    path: path,
    query: query,
    body: raw['body'],
    hasBody: raw.containsKey('body'),
    code: _optionalString(raw, 'code'),
    message: _optionalString(raw, 'message'),
    meta: meta,
  );
  frame.validate();
  return frame;
}

final class DecodedFrames {
  const DecodedFrames(this.frames, this.rest);

  final List<Frame> frames;
  final Uint8List rest;
}

DecodedFrames decodeStream(Uint8List buffer) {
  final frames = <Frame>[];
  final view = ByteData.sublistView(buffer);
  var offset = 0;
  while (buffer.length - offset >= lengthPrefixBytes) {
    final length = view.getUint32(offset, Endian.big);
    if (length > maxFrameBytes) {
      throw FrameException(
        'declared frame length $length is over the $maxFrameBytes limit',
      );
    }
    final start = offset + lengthPrefixBytes;
    if (buffer.length - start < length) break;
    frames.add(decodeFrame(Uint8List.sublistView(buffer, start, start + length)));
    offset = start + length;
  }
  return DecodedFrames(
    List.unmodifiable(frames),
    Uint8List.fromList(buffer.sublist(offset)),
  );
}

final class Correlator {
  Correlator([this.prefix = '']);

  final String prefix;
  int _next = 0;

  String take() {
    _next += 1;
    return '$prefix$_next';
  }
}

int _requiredInt(Map<String, Object?> raw, String name) {
  final value = raw[name];
  if (!raw.containsKey(name) || value is! int) {
    throw FrameException('$name has the wrong type');
  }
  return value;
}

String _requiredString(Map<String, Object?> raw, String name) {
  final value = raw[name];
  if (!raw.containsKey(name) || value is! String) {
    throw FrameException('$name has the wrong type');
  }
  return value;
}

String? _optionalString(Map<String, Object?> raw, String name) {
  if (!raw.containsKey(name)) return null;
  final value = raw[name];
  if (value is! String) throw FrameException('$name has the wrong type');
  return value;
}

bool _isJsonValue(Object? value) {
  if (value == null || value is bool || value is int) return true;
  if (value is double) return value.isFinite;
  if (value is String) return _isUnicodeScalarString(value);
  if (value is List) return value.every(_isJsonValue);
  if (value is Map) {
    return value.entries.every(
      (entry) =>
          entry.key is String &&
          _isUnicodeScalarString(entry.key as String) &&
          _isJsonValue(entry.value),
    );
  }
  return false;
}

bool _isUnicodeScalarString(String value) {
  final units = value.codeUnits;
  for (var index = 0; index < units.length; index += 1) {
    final unit = units[index];
    if (unit >= 0xd800 && unit <= 0xdbff) {
      if (index + 1 >= units.length) return false;
      final next = units[++index];
      if (next < 0xdc00 || next > 0xdfff) return false;
    } else if (unit >= 0xdc00 && unit <= 0xdfff) {
      return false;
    }
  }
  return true;
}
